### PR TITLE
[WebProfilerBundle] check for <title> instead of </body>

### PR DESCRIPTION
**Pull request by MatTheCat**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #20339 |
| License | MIT |
| Doc PR | - |

Valid HTML doesn't need `body` end tag but it need `title` element. As such we check for `<title>` instead of `</body>` before inserting the toolbar.

BC break will happen if response contains no `title` element. Dunno if it's a big deal.

### CODE REVIEW

**Comment by wouterj** [comment]
> BC break will happen if response contains no title element. Dunno if it's a big deal.

This can quite easily be avoided by a BC layer: If there is no `<title>` tag, check for `</body>` and trigger an `@trigger_error(..., E_USER_DEPRECATED);` error indicating that this is deprecated and you have to have a title tag.

**Review by wouterj — APPROVED** [review]

**Comment by MatTheCat** [comment]
Thanks. Do tests need to reflect that?

**Comment by fabpot** [comment]
The goal is to inject the toolbar at the bottom of the file. So, we only need to identify if the file is HTML (via `<title>` or `</body>`) but always inject it at the bottom. WDYT?

**Comment by MatTheCat** [comment]
Do you mean injecting the toolbar DOM or scripts generating it?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig [hunk 1]
```diff
@@ -1,13 +1,15 @@
-<div id="sfwdt{{ token }}" class="sf-toolbar" style="display: none"></div>
 {% include '@WebProfiler/Profiler/base_js.html.twig' %}
 <script>/*<![CDATA[*/
-    (function () {
+    document.addEventListener('DOMContentLoaded', function () {
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: No need to care about old browsers here because this is a tool for developers (in the profiler for example we use flexbox that it's not supported in old browsers).

**Comment by nicolas-grekas** [comment]
This is a new feature to me, should be on master.
Also, I don't see the need for the deprecation.
Detecting both body, title (and even html) tags looks fine to me.

**Comment by MatTheCat** [comment]
How is checking `<title>` instead of `</body>` a new feature? We're talking about a false negative here, seems to me it's a bugfix.

Checking only for `</body>` is erroneous so we check it for the sake of BC only.

**Comment by nicolas-grekas** [comment]
Adding a deprecation in 2.7 is the code smell that makes this more visible: we *never* add new deprecations to any bug-fix-only bracnches.

**Comment by MatTheCat** [comment]
Well let's remove deprecation on 2.7 then.

**Comment by MatTheCat** [comment]
Does that seem ok? If so I'll fix the tests.

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 2]
```diff
@@ -112,21 +112,30 @@ public function onKernelResponse(FilterResponseEvent $event)
...
-                    'position' => $this->position,
-                    'excluded_ajax_paths' => $this->excludedAjaxPaths,
-                    'token' => $response->headers->get('X-Debug-Token'),
-                    'request' => $request,
-                )
-            ))."\n";
-            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
-            $response->setContent($content);
+        $canInjectToolbar = false !== stripos($content, '<title>');
+        $injectToolbarAt = strripos($content, '</body>');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: I propose this:
```
$injectToolbarAt = strripos($content, '</body>') ?: strripos($content, '</html>');
$canInjectToolbar = false !== $injectToolbarAt || false !== stripos($content, '<title>');
```
then below:
```
$content = substr_replace($content, $toolbar, $injectToolbarAt ?: strlen($content), 0);
```

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Yep. In addition stripos/strripos cannot return zero on valid HTML so I think we can simplify second line like this

```php
$canInjectToolbar = $injectToolbarAt || stripos($content, '<title>');
```

Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: lets' go :)

**Comment by MatTheCat** [comment]
Should I add tests for checking script injection when there is no `</body>` or `</html>` tag?

**Comment by nicolas-grekas** [comment]
would be great, since that's how your started this PR

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig (hunk 1 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: @javiereguiluz the toolbar itself is injected in the page being displayed, so it is good to avoid breaking the page in older browsers (as it would make it impossible to test your site properly in older browsers).
This is why we applied bug fixes regarding the toolbar in IE8 in the past for instance (while not supporting IE8 in the profiler pages, as these can be viewed in a modern browser even when testing the page itself in an old browser)

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 3]
```diff
@@ -112,21 +112,30 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+
+        $canInjectToolbar |= false !== $injectToolbarAt;
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: using a binary operator here looks wrong to me, as ``$canInjectToolbar`` is an integer after than, not a boolean.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig [hunk 4]
```diff
@@ -1,13 +1,15 @@
-<div id="sfwdt{{ token }}" class="sf-toolbar" style="display: none"></div>
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: IMO, you should revert this change. Changing the way this div is added is not necessary for your bugfix

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 5]
```diff
@@ -112,21 +112,30 @@ public function onKernelResponse(FilterResponseEvent $event)
...
-                array(
-                    'position' => $this->position,
-                    'excluded_ajax_paths' => $this->excludedAjaxPaths,
-                    'token' => $response->headers->get('X-Debug-Token'),
-                    'request' => $request,
-                )
-            ))."\n";
-            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
-            $response->setContent($content);
+        $canInjectToolbar = false !== stripos($content, '<title>');
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: do we actually need to test for ``<title>`` if we add support for ``</html>`` for pages not having ``</body>`` ? Is it valid to omit the ``</html>`` ?

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 6]
```diff
@@ -112,21 +112,24 @@ public function onKernelResponse(FilterResponseEvent $event)
...
-                    'position' => $this->position,
-                    'excluded_ajax_paths' => $this->excludedAjaxPaths,
-                    'token' => $response->headers->get('X-Debug-Token'),
-                    'request' => $request,
-                )
-            ))."\n";
-            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
-            $response->setContent($content);
+        $injectToolbarAt = strripos($content, '</body>') ?: strripos($content, '</html>');
+        $canInjectToolbar = $injectToolbarAt || stripos($content, '<title>');
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: having a ``<title>`` tag does not ensure you return a full HTML page AFAIK. you could be sending a partial containing the ``<head>`` only, in which case the toolbar should not be injected.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig (hunk 4 above)
Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: You're right, it was only necessary when I injected the scripts before `<title>`. That being said does this matter?

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php (hunk 6 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Currently, we err on the side of not injecting the toolbar when it could be (but you can still access the data by looking for recent profiles in the profiler). Your PR allows to break the response in some case by injecting the toolbar when it should not.

Is it possible to write a valid HTML full page without a ``</html>`` (or without a ``<html`` ? If no, these should be used for the detecting instead IMO (``</html>`` is already used by your change btw)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig (hunk 4 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: As you want to have it merged as a bug fix in 2.7, it should avoid doing any unnecessary changes (which have a risk of having a negative impact while providing no value)

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig]: I see, I'll revert this

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php (hunk 6 above)
Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: `<!DOCTYPE><title>title</title>` is a valid HTML page. But what do you call a "partial"? Checks exist against subrequests.

Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: you can have a normal HTTP request returning partial HTML (generally done through XHR though, but detecting XHR requests cannot be done. ``isXMLHttpRequest`` detects a special header added by jQuery and a few other major frameworks, but it may not be there when using the native APIs directly).

Another case of partial is the response to an ESI request (which is **not** a subrequest, especially when the ESI is resolved outside Symfony)

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Only difference is now we're checking for `<title>`,  `</body>` and `</html>` there's more chance to add the toolbar in a partial.

Comment (xabbuh) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: ESI requests are something commonly used. Imo the risk isn't just a minor edge case.

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: ESI requests including `<title>`, `</body>` or `</html>`?

Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: something including ``</body>`` or ``</html>`` is safe, as we want to inject things at the end of the page, and the detection precisely relies on identifying the injection place. The issue IMO is checking for ``<title>`` to detect that we should inject it, and then injecting at the end of the response because we cannot detect the right injection place.

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Before `<title>` would be a right injection place.

Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: no it is not, as we want to inject the toolbar itself

Comment (MatTheCat) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Toolbar can be injected with JavaScript

**Comment by MatTheCat** [comment]
Seems I'd better open a PR agains master checking for `<title>` only?

**Comment by fabpot** [comment]
Reading the conversation here, I err on the side of a won't fix for the same reasons as explained by @stof and @xabbuh. We don't want to inject the toolbar if we are not really sure that this is a full HTML page. That's the case now as a `</body>` cannot be part of an HTML fragment. `<title>`
can be used in a fragment in which case it would incorrectly inject the toolbar.

**Comment by MatTheCat** [comment]
How couldn't `</body>` be part of an HTML fragment?

**Comment by fabpot** [comment]
Because `</body>` can only be part of a full HTML page, where we want to inject the toolbar.

**Comment by stof** [comment]
@MatTheCat we want the toolbar to be injected at the end of the body. If we have the end of the body in our response, we know we want to inject it there. And making an ESI request wrapping the whole ``<body>`` of the page (leaving only ``<html>`` and the head in the main response) would be totally insane anyway (which is the only way I can think of to generate some response with a closing body tag without it being the end of the page). And in such case, having the toolbar of the request generating the body would not be that confusing for the user (even less considering the weird thing they were already doing with ESI)

**Comment by MatTheCat** [comment]
The toolbar could be injected nearly anywhere: it's just scripts afterall. If a response is HTML and contains `<title>` I can't see why it would be a problem to inject it before?

**Comment by fabpot** [comment]
If you are using ESI, you don't want to have several toolbars.

**Comment by MatTheCat** [comment]
There should be only one `<title>` so there should be only one toolbar.

**Comment by xabbuh** [comment]
There will be a second toolbar being injected in the request that returns the closing `body` element.

**Comment by MatTheCat** [comment]
Not if we check only for `<title>`. That would be a BC break so I proposed to open a new PR on master.

**Comment by fabpot** [comment]
As 3 core team members think this is a bad idea, I'm closing this PR as a won't fix. Thanks for your understanding.

**Comment by MatTheCat** [comment]
What about the fix on master?
